### PR TITLE
Iptv matrix

### PR DESCRIPTION
**Pull request by phunkyfish**
v4.7.0
- Added: Support for IPTV Streams configured on E2 device (no timeshifting)
- Added: Reload instead of reconnecting when channel/group changes are detected
- Added: Use truly unique IDs for channels so EPG changes are correctly reflected
- Fixed: Only get drive space for devices that have an HDD
- Fixed: use correct function to lookup group when adding
- Added: update README.md to show appveyor/travis badges per branch
- Added: Update OSX build script
- Added: update badge status for travis/appveyor
- Added: add copyright notices to files
- Fixed: Fix default path for genre text mapping file

### CODE REVIEW

**Comment by phunkyfish** [comment]
@ksooo this is ready for review when you have the time.

**Comment by AlwinEsch** [comment]
That's a good and great rework. Hopefully, PVR will soon be switched to the C ++ interface.

If you have some ideas for rework. add it here :smile:: https://github.com/xbmc/xbmc/pull/16485
Maybe you can already prepare for the addon in part (if possible).

**Comment by phunkyfish** [comment]
Ty, yes, that’s the plan for the next PRs for vuplus and iptvsimple.

File: src/Enigma2.cpp [hunk 1]
```diff
@@ -484,6 +484,11 @@ const std::string Enigma2::GetLiveStreamURL(const PVR_CHANNEL &channelinfo)
   return m_channels.GetChannel(channelinfo.iUniqueId)->GetStreamURL();
 }
 
+bool Enigma2::IsIptvStream(const PVR_CHANNEL& channelinfo)
```
Comment (ksooo) [inline src/Enigma2.cpp]: make member function `const` ?

File: src/client.cpp [hunk 2]
```diff
@@ -357,13 +357,25 @@ PVR_ERROR GetChannelStreamProperties(const PVR_CHANNEL* channel, PVR_NAMED_VALUE
...
 
-  const std::string strStreamProgramNumber = std::to_string(enigma->GetChannelStreamProgramNumber(*channel));
+  *iPropertiesCount = 0;
 
-  Logger::Log(LEVEL_NOTICE, "%s - for channel: %s, set Stream Program Number to %s - %s", __FUNCTION__, channel->strChannelName, strStreamProgramNumber.c_str(), enigma->GetLiveStreamURL(*channel).c_str());
+  if (enigma->IsIptvStream(*channel))
+  {
+    strncpy(properties[0].strName, PVR_STREAM_PROPERTY_STREAMURL, sizeof(properties[0].strName) - 1);
+    strncpy(properties[0].strValue, enigma->GetLiveStreamURL(*channel).c_str(), sizeof(properties[0].strValue) - 1);
+    *iPropertiesCount += 1;
```
Comment (ksooo) [inline src/client.cpp]: `*iPropertiesCount++` would be shorter.

File: src/client.cpp [hunk 3]
```diff
@@ -357,13 +357,25 @@ PVR_ERROR GetChannelStreamProperties(const PVR_CHANNEL* channel, PVR_NAMED_VALUE
...
+    const std::string strStreamProgramNumber = std::to_string(enigma->GetChannelStreamProgramNumber(*channel));
 
-  strncpy(properties[0].strName, "program", sizeof(properties[0].strName) - 1);
-  strncpy(properties[0].strValue, strStreamProgramNumber.c_str(), sizeof(properties[0].strValue) - 1);
-  *iPropertiesCount = 1;
+    Logger::Log(LEVEL_NOTICE, "%s - for channel: %s, set Stream Program Number to %s - %s", __FUNCTION__, channel->strChannelName, strStreamProgramNumber.c_str(), enigma->GetLiveStreamURL(*channel).c_str());
+
+    strncpy(properties[0].strName, "program", sizeof(properties[0].strName) - 1);
+    strncpy(properties[0].strValue, strStreamProgramNumber.c_str(), sizeof(properties[0].strValue) - 1);
+    *iPropertiesCount += 1;
```
Comment (ksooo) [inline src/client.cpp]: `*iPropertiesCount++` would be shorter.

File: src/enigma2/data/Channel.cpp [hunk 4]
```diff
@@ -65,23 +65,37 @@ bool Channel::UpdateFrom(TiXmlElement* channelNode)
...
-    commonServiceReference.c_str()
-  );
+  if (!m_isIptvStream)
+  {
+    m_streamURL = StringUtils::Format(
+      "http%s://%s%s:%d/%s",
+      Settings::GetInstance().UseSecureConnectionStream() ? "s" : "",
+      Settings::GetInstance().UseLoginStream() ? StringUtils::Format("%s:%s@", Settings::GetInstance().GetUsername().c_str(), Settings::GetInstance().GetPassword().c_str()).c_str() : "",
+      Settings::GetInstance().GetHostname().c_str(),
+      Settings::GetInstance().GetStreamPortNum(),
```
Comment (ksooo) [inline src/enigma2/data/Channel.cpp]: Introduce variable for the value of `Settings::GetInstance()`? You are calling it  5 times in a row here and at least one more time in `Channel::UpdateFrom`.

File: src/enigma2/Channels.cpp [hunk 5]
```diff
@@ -48,7 +48,15 @@ int Channels::GetChannelUniqueId(const std::string& channelServiceReference)
 
 std::shared_ptr<Channel> Channels::GetChannel(int uniqueId)
 {
-  return m_channels.at(uniqueId - 1);
+  std::shared_ptr<Channel> channel = nullptr;
```
Comment (ksooo) [inline src/enigma2/Channels.cpp]: no need for `= nullptr`. `shared_ptr` default ctor will do this automatically.

How about rewriting this method:
```c++
  auto channelPair = m_channelsUniqueIdMap.find(uniqueId);
  if (channelPair != m_channelsUniqueIdMap.end())
    return channelPair->second;

 return {};
```

File: src/enigma2/Channels.cpp [hunk 6]
```diff
@@ -359,4 +369,18 @@ ChannelsChangeState Channels::CheckForChannelAndGroupChanges(enigma2::ChannelGro
     return ChannelsChangeState::CHANNEL_GROUPS_CHANGED;
 
   return ChannelsChangeState::NO_CHANGE;
+}
+
+int Channels::GenerateChannelUniqueId(const std::string channelName, const std::string extendedServiceReference)
```
Comment (ksooo) [inline src/enigma2/Channels.cpp]: pass string params as references, not by value

File: src/enigma2/Channels.h [hunk 7]
```diff
@@ -66,13 +66,16 @@ namespace enigma2
     ChannelsChangeState CheckForChannelAndGroupChanges(enigma2::ChannelGroups& latestChannelGroups, enigma2::Channels& latestChannels);
 
   private:
+    static int GenerateChannelUniqueId(const std::string channelName, const std::string extendedServiceReference);
```
Comment (ksooo) [inline src/enigma2/Channels.h]: pass string params as references, not by value

Comment (ksooo) [reply src/enigma2/Channels.h]: A good alternative for static private methods are local functions in the cpp file which are kept in an unnamed namespace. This way you don't need to expose internal stuff to header files.

File: pvr.vuplus/addon.xml.in [hunk 8]
```diff
@@ -175,6 +175,18 @@
     <disclaimer lang="zh_TW">這是測試版軟體！其原創作者並無法對於以下情況負責，包含：錄影失敗，不正確的定時設定，多餘時數，或任何產生的其它不良影響...</disclaimer>
     <platform>@PLATFORM@</platform>
     <news>
+v4.6.0
```
Comment (ksooo) [inline pvr.vuplus/addon.xml.in]: This should read v4.7.0, actual changelog for v4.6.0 is missing and more important, you forgot to increase the addon version number in line 4 of this file.

**Review by ksooo — COMMENTED** [review]
some remarks/questions

File: src/Enigma2.cpp (hunk 1 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: The underlying call to `m_channels` is not const so this won't work.

File: src/client.cpp (hunk 2 above)
Comment (phunkyfish) [reply src/client.cpp]: `(*iPropertiesCount)++` is shorter but uglier. `*iPropertiesCount++` will give a warning. Hence I went with `+= 1`

File: pvr.vuplus/addon.xml.in (hunk 8 above)
Comment (phunkyfish) [reply pvr.vuplus/addon.xml.in]: Ya, I did this in iptvsimple too since the API bump :(

File: src/client.cpp (hunk 2 above)
Comment (phunkyfish) [reply src/client.cpp]: In practice which format would be most common to see? Then I'll just start doing that ;)

Comment (phunkyfish) [reply src/client.cpp]: Ok, just went with `(*iPropertiesCount)++`

File: src/enigma2/data/Channel.cpp (hunk 4 above)
Comment (phunkyfish) [reply src/enigma2/data/Channel.cpp]: Ya, I have a messy include hierarchy to fix up here to do this. I think I will save this until a later refactor if that's ok.

File: src/enigma2/Channels.cpp (hunk 5 above)
Comment (phunkyfish) [reply src/enigma2/Channels.cpp]: Done

File: src/enigma2/Channels.cpp (hunk 6 above)
Comment (phunkyfish) [reply src/enigma2/Channels.cpp]: Done

File: src/enigma2/Channels.h (hunk 7 above)
Comment (phunkyfish) [reply src/enigma2/Channels.h]: Done

File: src/enigma2/data/Channel.cpp (hunk 4 above)
Comment (ksooo) [reply src/enigma2/data/Channel.cpp]: I don't see how this is related to include hierarchy, but it's up to you.

File: src/Enigma2.cpp (hunk 1 above)
Comment (ksooo) [reply src/Enigma2.cpp]: Can this or underlying function made be logically const? => make a member mutable

Comment (phunkyfish) [reply src/Enigma2.cpp]: Sure, I guess the vector can be mutable. Let me check if `GetChannel` in general can be.

File: src/enigma2/data/Channel.cpp (hunk 4 above)
Comment (phunkyfish) [reply src/enigma2/data/Channel.cpp]: Fair, it’s easy local to that function. I was trying to do it as a class member.

Comment (ksooo) [reply src/enigma2/data/Channel.cpp]: I ment local to the function.

File: src/Enigma2.cpp (hunk 1 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: Done

**Comment by phunkyfish** [comment]
Ok, should be ready. Same changes made to Leia PR.

**Review by ksooo — APPROVED** [review]

**Comment by phunkyfish** [comment]
Thanks!

**Pull request merged**